src/open_petro_elastic/material/span_wagner/carbon_dioxide.py

```diff
@@ -0,0 +1,420 @@
+
+    bounds[below_triple, 1] = saturated_vapor_density(
+        absolute_temperature[below_triple]
+    )
+    if force_vapor is True:
+        bounds[below_critical, 1] = saturated_vapor_density(
+            absolute_temperature[below_critical]
+        )
+    elif force_vapor is False:
+        bounds[below_critical, 0] = saturated_liquid_density(
+            absolute_temperature[below_critical]
+        )
+    else:  # force_vapor == 'auto'
+        below_vapor_pressure = pressure < vapor_pressure(absolute_temperature)
+        is_vapor = below_critical & below_vapor_pressure
+        bounds[is_vapor, 1] = saturated_vapor_density(absolute_temperature[is_vapor])
+        is_liquid = below_critical & ~below_vapor_pressure
+        bounds[is_liquid, 0] = saturated_liquid_density(absolute_temperature[is_liquid])
+    return bounds
+
+
+def _find_initial_density_values(bounds, absolute_temperature, pressure):
+    """
+    Finds approximate density values for the provided temperature(s) and pressure(s). The result is only intended to be
+    used by array_carbon_dioxide_density.
+    """
+    from scipy.interpolate import RegularGridInterpolator
+
+    temps = np.geomspace(
+        np.min(absolute_temperature) * 0.99, np.max(absolute_temperature) * 1.01, 41
+    )
+    press = np.geomspace(
+        np.min(pressure) * 0.99, np.max(absolute_temperature) * 1.01, 41
+    )
+    tt, pp = np.meshgrid(temps, press, indexing="ij")
+    densi = carbon_dioxide_density(
+        tt.flatten(), pp.flatten(), force_vapor="auto", raise_error=False
+    ).reshape(temps.size, press.size)
+    rgi = RegularGridInterpolator(np.array((temps, press)), densi, method="linear")
+    iv = rgi(np.array((absolute_temperature, pressure)).T)
+    oob = (iv < bounds[:, 0]) | (iv > bounds[:, 1]) | np.isnan(iv)
+    iv[oob] = np.mean(bounds[oob], axis=1)
+    return iv
+
+
+def array_carbon_dioxide_density(absolute_temperature, pressure, force_vapor):
+    """
+    Alternative implementation of a vectorized carbon dioxide density function. Implemented primarily for demonstration
+    purposes. For large arrays, a look-up-table approach should be preferred.
+
+    Utilizes scipy.optimize.newton, which is the only root-finding method of scipy that supports a vectorized functions.
+
+    For argument documentation, see carbon_dioxide_density.
+    """
+    absolute_temperature = np.asarray(absolute_temperature)
+    pressure = np.asarray(pressure)
+    bounds = _determine_density_bounds(absolute_temperature, pressure, force_vapor)
+    iv = _find_initial_density_values(bounds, absolute_temperature, pressure)
+    opt = scipy.optimize.newton(
+        lambda x: carbon_dioxide_pressure(absolute_temperature, x) - pressure,
+        x0=iv,
+        maxiter=10,
+        full_output=True,
+    )
+    # scipy.optimize.newton may not always converge. We need to determine which of the elements of the solution are
```

Comment:
It would be good to reformat comments so that they do not wrap in git, say 120 character limit.